### ISSUE TITLE
MySQL defaulting to localhost's external IP when trying to connect to a remote DB server

### ISSUE THREAD

**Issue by archonxgames** [jaredwray/keyv#516]
**Describe the bug**
When I try to connect to a remote database using keyv mysql package, I get an Access denied error. Upon inspection, a connection attempt was made not to the remote server host but to the local host's external IP address. The connection string is correct and there were no other changes made to the code besides the connection attempt.

**How To Reproduce (best to provide workable code or tests!)**
Localhost's External IP (sample): 111.111.111.111
Remote DB Host IP: 12.34.567.890

```js
const Keyv = require('keyv');
const connString = 'mysql://user:password@12.34.567.890:3306/db_name'
console.log(connString)

const keyv = new Keyv(connString)
keyv.on('error', err => console.error('Keyv connection error:', err))
```

Error:
```
  code: 'ER_ACCESS_DENIED_ERROR',
  errno: 1045,
  sql: undefined,
  sqlState: '28000',
  sqlMessage: "Access denied for user 'user'@'111.111.111.111' (using password: YES)"
```

**Comment by jaredwray** [comment]
Hi @archonxgames - I am not able to duplicate this issue. Is there a way to setup this environment to validate this? Do you have a docker scenario that we can use as a test environment?

**Comment by jaredwray** [comment]
closing due to no response and unable to reproduce

**Comment by archonxgames** [comment]
Apologies for the late reply, unfortunately my technical acumen with regards to Docker is completely squat, though I encountered this issue while trying to use Keyv together with Discord.js. Other than that, I unfortunately have no other leads due to my relative inexperience (Im currently learning multiple libraries together for the first time)

**Comment by jaredwray** [comment]
ok thanks.

**Comment by skywiind** [comment]
Hi, I'm also having this issue where keyv is defaulting to my external IPv4. 
I essentially had the same code they did, where I wrote the connection string to a string and then passed that into the keyv constructor. I confirmed that the port on the IP I'm trying to connect to is open and I am able to ping the IP no problem. That section of my code looks like this: 
```
const Keyv = require("keyv");

const dbUser = "user";
const dbPass = "password";
const dbIP     = "<serverIP>:3306";
const dbPath = "mysql";

const dbo = new Keyv(`mysql://${dbUser}:${dbPass}@${dbIP}/${dbPath}`);
// "mysql://user:pass@<serverIP>:3306/mysql"
dbo.on("error", err => console.error("Keyv connection error:", err));
```

At this point, it throws this error, just like the other person:
```
Error: Access denied for user 'user'@'<myexternalIP>' (using password: YES)
    at PromisePool.execute (D:\Documents\C\node_modules\mysql2\promise.js:369:22)
    at D:\Documents\C\node_modules\@keyv\mysql\src\index.js:32:42
    at D:\Documents\C\node_modules\@keyv\mysql\src\index.js:41:19
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  code: 'ER_ACCESS_DENIED_ERROR',
  errno: 1045,
  sql: undefined,
  sqlState: '28000',
  sqlMessage: "Access denied for user 'user'@'<myexternalIP>' (using password: YES)"
}
```
I followed the stack trace and I didn't see anything glaringly wrong (not that I have the knowledge to spot such problems). Like the previous person, I made sure everything is up-to-date.

It's worth noting that I am able to connect to this database in any mySQL client with the exact same details. I have tested opening connections to mySQL servers with hostnames and bare IP addresses and had no issues. Logically, this just looks like some sort of configuration problem with the server itself or with my environment. However, this still persists when I run the code on a machine outside my home network. 

To summarize: 
- Error persists on one IP. Other IPs and hostnames work fine.
- Error occurs on multiple machines, with the erroneous IP matching the IP of the respective machine.
- IP and port are confirmed open
- mySQL clients can connect to the database

Let me know if there's any additional information I can provide to help with diagnosing this issue. If you need me to set up an environment that you can access to get a better look at what's going on, I can _probably_ have that up and running in a couple days. Unfortunately, like the previous person, I'm also not crazy familiar with debugging when the other person isn't working in the same environment.

**Comment by afreemanio** [comment]
I ran into this one today as well. This is really strange - using keyv alongside prisma in the same repo, and the same url works for Prisma but not for KeyV. Both connecting to MySQL through Docker. It rewrites the host 'host.docker.internal' as well as '172.17.0.1' (which route to the docker host) to the IP of the container itself. I have never seen such a thing before...

**Comment by jaredwray** [comment]
do you have a docker setup I can test with?

**Comment by artn** [comment]
Same issue.

it seems that keyv storage driver uses https://github.com/sidorares/node-mysql2 and [its document](https://github.com/sidorares/node-mysql2#using-connection-pools) indicate that  connection parameter is not string but object like below.
```
async function main() {
  // get the client
  const mysql = require('mysql2');
  // create the pool
  const pool = mysql.createPool({host:'localhost', user: 'root', database: 'test'});
  // now get a Promise wrapped instance of that pool
  const promisePool = pool.promise();
  // query database using promises
  const [rows,fields] = await promisePool.query("SELECT 1");
```

**Comment by jaredwray** [comment]
@artn - Hav you been able to confirm that this would fix the issue if we did it as an object? I have not been able to get to this yet and planned to look at this more next week.

**Comment by jaredwray** [comment]
@artn - just did a change to support the config better which looks like it could cause issues. https://github.com/jaredwray/keyv/pull/791 -- merged but wont be published for a week or so. 

@afreemanio - This looks like it solves the issue we are seeing so please test but I am going to close this out as it will be live in the next couple weeks. 

@skywiind @archonxgames FYI that this fix is coming.
